In [51]:
import pandas as pd
import numpy as np
import re

In [52]:
df=pd.read_csv("/content/drive/MyDrive/Resume_dataset.csv")

In [53]:
df.head()

,Resume,Category
0,Python Developer Python Developer Philadelphia...,Python_Developer
1,Python Developer Python Developer Python Devel...,Python_Developer
2,R&D Engineer R&D Engineer R&D Engineer - Nokia...,Python_Developer
3,Sr. Full Stack Developer Sr. Full Stack Develo...,Python_Developer
4,Sr. Full Stack Python Developer Sr. Full Stack...,Python_Developer


In [54]:
df["Category"].value_counts()

,count
Category,
Python_Developer,1278
Java_Developer,1242
Web_Developer,984
Database_Administrator,920
Security_Analyst,876
Systems_Administrator,776
Project_manager,722
Front_End_Developer,530
Network_Administrator,469


In [55]:
df.columns

Index(['Resume', 'Category'], dtype='object')

In [56]:
df.shape

(8234, 2)

In [57]:
X=df["Resume"]
y=df["Category"]

In [58]:
from sklearn.preprocessing import LabelEncoder
from sklearn.svm import LinearSVC
from sklearn.feature_extraction.text import TfidfVectorizer #TF-IDF

In [59]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import (accuracy_score,classification_report,confusion_matrix)

In [60]:
def preprocess(text):
  text=text.lower()
  text=re.sub(r'[^A-Za-z0-9 ]',"",text)
  text = re.sub(r'\s+', ' ', text)
  return text.strip()


In [61]:
X

,Resume
0,Python Developer Python Developer Philadelphia...
1,Python Developer Python Developer Python Devel...
2,R&D Engineer R&D Engineer R&D Engineer - Nokia...
3,Sr. Full Stack Developer Sr. Full Stack Develo...
4,Sr. Full Stack Python Developer Sr. Full Stack...
...,...
8229,Oracle database administrator Oracle database ...
8230,"Job Seeker Columbus, OH I am responsible and d..."
8231,Database Administrator Database Administrator ...
8232,Business Intelligence Engineer Business Intell...


In [62]:
X=X.apply(preprocess)

In [63]:
X

,Resume
0,python developer python developer philadelphia...
1,python developer python developer python devel...
2,rd engineer rd engineer rd engineer nokia solu...
3,sr full stack developer sr full stack develope...
4,sr full stack python developer sr full stack p...
...,...
8229,oracle database administrator oracle database ...
8230,job seeker columbus oh i am responsible and de...
8231,database administrator database administrator ...
8232,business intelligence engineer business intell...


In [64]:
le=LabelEncoder()
y=le.fit_transform(y)

In [65]:
le.classes_

array(['Database_Administrator', 'Front_End_Developer', 'Java_Developer',
       'Network_Administrator', 'Project_manager', 'Python_Developer',
       'Security_Analyst', 'Software_Developer', 'Systems_Administrator',
       'Web_Developer'], dtype=object)

In [66]:
y

array([5, 5, 5, ..., 0, 0, 0])

In [67]:
inverse_mapping = dict(zip(range(len(le.classes_)), le.classes_))
print(inverse_mapping)

{0: 'Database_Administrator', 1: 'Front_End_Developer', 2: 'Java_Developer', 3: 'Network_Administrator', 4: 'Project_manager', 5: 'Python_Developer', 6: 'Security_Analyst', 7: 'Software_Developer', 8: 'Systems_Administrator', 9: 'Web_Developer'}


In [68]:
tfidf = TfidfVectorizer(
    max_features=5000,
    ngram_range=(1,2)
)

X_vec= tfidf.fit_transform(X)

In [69]:
X_vec.shape

(8234, 5000)

In [70]:
X.shape

(8234,)

In [71]:
X_train, X_test, y_train, y_test = train_test_split(
    X_vec,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [72]:
model = LinearSVC()

model.fit(X_train, y_train)

LinearSVC()

In [73]:
pred_linear_svc= model.predict(X_test)

In [74]:
print("Accuracy")

print(
    accuracy_score(y_test, pred_linear_svc)
)

Accuracy
0.9344262295081968


In [75]:
print(
    classification_report(
        y_test,
        pred_linear_svc,
        target_names=le.classes_
    )
)

                        precision    recall  f1-score   support

Database_Administrator       0.97      0.96      0.96       184
   Front_End_Developer       0.83      0.80      0.81       106
        Java_Developer       0.97      0.99      0.98       249
 Network_Administrator       0.91      0.79      0.85        94
       Project_manager       0.96      0.98      0.97       144
      Python_Developer       0.99      0.99      0.99       256
      Security_Analyst       0.98      0.97      0.97       175
    Software_Developer       0.81      0.79      0.80        87
 Systems_Administrator       0.88      0.95      0.91       155
         Web_Developer       0.90      0.89      0.90       197

              accuracy                           0.93      1647
             macro avg       0.92      0.91      0.91      1647
          weighted avg       0.93      0.93      0.93      1647



In [76]:
cm = confusion_matrix(
    y_test,
    pred_linear_svc
)

print(cm)

[[177   0   0   0   5   0   1   0   1   0]
 [  0  85   1   0   0   0   0   8   0  12]
 [  1   0 247   0   0   0   0   1   0   0]
 [  1   0   1  74   0   1   1   0  15   1]
 [  0   0   0   0 141   0   0   1   2   0]
 [  0   0   2   0   0 254   0   0   0   0]
 [  1   0   0   2   1   0 169   0   2   0]
 [  0  10   1   0   0   0   0  69   0   7]
 [  2   0   0   5   0   0   1   0 147   0]
 [  1   8   3   0   0   1   1   6   1 176]]


In [77]:
resume = """
Python
Pandas
NumPy
Machine Learning
TensorFlow
Deep Learning
SQL
"""

resume = preprocess(resume)

resume_vector = tfidf.transform([resume])

prediction = model.predict(
    resume_vector
)

print(
    le.inverse_transform(prediction)
)

['Database_Administrator']


In [78]:
resume_2 = """
html
css
Javascript
react
"""

resume = preprocess(resume_2)

resume_vector = tfidf.transform([resume_2])

prediction = model.predict(
    resume_vector
)

print(
    le.inverse_transform(prediction)
)

['Software_Developer']


In [82]:
import sys
!{sys.executable} -m pip install gensim
from gensim.models import Word2Vec

In [84]:
from sklearn.naive_bayes import GaussianNB

In [85]:
X

,Resume
0,python developer python developer philadelphia...
1,python developer python developer python devel...
2,rd engineer rd engineer rd engineer nokia solu...
3,sr full stack developer sr full stack develope...
4,sr full stack python developer sr full stack p...
...,...
8229,oracle database administrator oracle database ...
8230,job seeker columbus oh i am responsible and de...
8231,database administrator database administrator ...
8232,business intelligence engineer business intell...


In [86]:
y

array([5, 5, 5, ..., 0, 0, 0])

In [88]:
X_tokenize = X.apply(lambda x: x.split())

In [89]:
X_tokenize

,Resume
0,"[python, developer, python, developer, philade..."
1,"[python, developer, python, developer, python,..."
2,"[rd, engineer, rd, engineer, rd, engineer, nok..."
3,"[sr, full, stack, developer, sr, full, stack, ..."
4,"[sr, full, stack, python, developer, sr, full,..."
...,...
8229,"[oracle, database, administrator, oracle, data..."
8230,"[job, seeker, columbus, oh, i, am, responsible..."
8231,"[database, administrator, database, administra..."
8232,"[business, intelligence, engineer, business, i..."


In [91]:
type (X_tokenize)

pandas.core.series.Series

In [92]:
sentences = X_tokenize.tolist()

In [93]:
w2v_model = Word2Vec(
    sentences=sentences,
    vector_size=100,
    window=5,
    min_count=2,
    workers=4,
    sg=1
)

In [94]:
def document_vector(words):

    word_vectors = []

    for word in words:

        if word in w2v_model.wv:
            word_vectors.append(w2v_model.wv[word])

    if len(word_vectors) == 0:

        return np.zeros(w2v_model.vector_size)

    return np.mean(word_vectors,axis=0)


In [95]:
X_vec = np.array([document_vector(tokens) for tokens in X_tokenize])

In [99]:
X_vec.shape

(8234, 100)

In [100]:
w2v_model.wv["python"]

array([-0.15522656, -0.202531  ,  0.23289332,  0.62675375, -0.36058852,
        0.2335136 , -0.16011351,  0.8867265 , -0.3158518 , -0.02778697,
       -0.44348863, -0.5635849 , -0.8340607 ,  0.50575864, -0.14545053,
       -0.3840475 , -0.5175126 , -0.00814353,  0.49732482, -0.23969038,
        0.167645  ,  0.18436252,  0.1949175 , -0.5445196 ,  0.08288397,
        0.04317072, -0.36994722,  0.24491064, -0.2706176 , -0.12483713,
        0.13284427, -0.2521828 ,  0.13966386,  0.04656135, -0.07267042,
       -0.10827111, -0.08115909, -0.77412105, -0.72948915,  0.29264364,
        0.44945744,  0.1700913 , -0.0656469 , -0.22716293,  0.4358578 ,
        0.3781103 , -0.59208834, -0.11390171, -0.06429312,  0.01502355,
       -0.4201609 , -0.23514265,  0.2585559 ,  0.33600298,  0.548554  ,
        0.00919041,  0.1565578 , -0.5983721 , -0.2560668 , -0.08293073,
        0.41414297, -0.01741499, -0.58200246,  0.5136614 ,  0.5126672 ,
        0.11198252, -0.17931876, -0.17319874, -0.189615  ,  0.63

In [101]:
w2v_model.wv.most_similar('python')

[('perl', 0.7467702031135559),
 ('beautifulsoup4', 0.7258800268173218),
 ('3527', 0.7207192182540894),
 ('erlang', 0.7205673456192017),
 ('scriptiong', 0.7163800597190857),
 ('python2', 0.7086529731750488),
 ('2730', 0.7034692168235779),
 ('python2x3x', 0.6991034746170044),
 ('lua', 0.697589099407196),
 ('2734', 0.6962343454360962)]

In [105]:
w2v_model.wv.most_similar('karnataka')

[('bengaluru', 0.9439300894737244),
 ('bangalore', 0.8687170743942261),
 ('ahmedabad', 0.8499826788902283),
 ('maharashtra', 0.841100811958313),
 ('tamil', 0.8377100825309753),
 ('nadu', 0.8256492018699646),
 ('chennai', 0.8248907327651978),
 ('uttar', 0.8242236375808716),
 ('chandigarh', 0.8231697678565979),
 ('pune', 0.8230646848678589)]

In [106]:
from sklearn.metrics.pairwise import cosine_similarity

python_vec = w2v_model.wv["python"]
perl_vec = w2v_model.wv["perl"]

cosine_similarity(
    [python_vec],
    [perl_vec]
)

array([[0.7467701]], dtype=float32)

In [107]:

X_train, X_test, y_train, y_test = train_test_split(
    X_vec,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


In [108]:
model = GaussianNB()

In [109]:
model.fit(
    X_train,
    y_train
)


GaussianNB()

In [110]:
y_pred_w2v = model.predict(
    X_test
)


In [111]:
print(
    "Accuracy :",
    accuracy_score(
        y_test,
        y_pred_w2v
    )
)

Accuracy : 0.7176684881602914


In [113]:
print(
    classification_report(
        y_test,
        y_pred_w2v,
        target_names=le.classes_
    )
)

print(
    confusion_matrix(
        y_test,
        y_pred_w2v
    )
)

                        precision    recall  f1-score   support

Database_Administrator       0.93      0.70      0.80       184
   Front_End_Developer       0.71      0.69      0.70       106
        Java_Developer       0.84      0.80      0.82       249
 Network_Administrator       0.52      0.46      0.49        94
       Project_manager       0.63      0.91      0.74       144
      Python_Developer       0.93      0.84      0.88       256
      Security_Analyst       0.95      0.70      0.81       175
    Software_Developer       0.30      0.38      0.33        87
 Systems_Administrator       0.65      0.83      0.73       155
         Web_Developer       0.51      0.55      0.53       197

              accuracy                           0.72      1647
             macro avg       0.70      0.69      0.68      1647
          weighted avg       0.75      0.72      0.72      1647

[[129   0   0   9  35   0   3   3   5   0]
 [  0  73   1   0   0   0   0  11   0  21]
 [  0   0 199  

In [114]:
from sklearn.linear_model import LogisticRegression

In [116]:
model_log=LogisticRegression()

In [117]:
model_log.fit(
    X_train,
    y_train
)

LogisticRegression()

In [118]:
y_pred_w2v_log = model_log.predict(
    X_test
)


In [119]:
print(
    "Accuracy :",
    accuracy_score(
        y_test,
        y_pred_w2v_log
    )
)

Accuracy : 0.8239222829386764


In [120]:
from gensim.models import FastText

In [122]:
X_tokenize

,Resume
0,"[python, developer, python, developer, philade..."
1,"[python, developer, python, developer, python,..."
2,"[rd, engineer, rd, engineer, rd, engineer, nok..."
3,"[sr, full, stack, developer, sr, full, stack, ..."
4,"[sr, full, stack, python, developer, sr, full,..."
...,...
8229,"[oracle, database, administrator, oracle, data..."
8230,"[job, seeker, columbus, oh, i, am, responsible..."
8231,"[database, administrator, database, administra..."
8232,"[business, intelligence, engineer, business, i..."


In [124]:
ft_model = FastText(
    sentences=X_tokenize,
    vector_size=100,
    window=5,
    min_count=2,
    workers=8,
    sg=1
)

In [125]:
def document_vector(words):

    vectors = []

    for word in words:

        vectors.append(ft_model.wv[word])

    return np.mean(vectors,axis=0)

In [126]:
X = np.array([document_vector(tokens)for tokens in X_tokenize])

In [127]:
print("X Shape :", X.shape)

X Shape : (8234, 100)


In [128]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [129]:
model = LogisticRegression(
    max_iter=2000
)

model.fit(
    X_train,
    y_train
)

# ==========================================
# Prediction
# ==========================================

pred = model.predict(
    X_test
)

# ==========================================
# Evaluation
# ==========================================

print(
    "Accuracy :",
    accuracy_score(
        y_test,
        pred
    )
)

print(
    classification_report(
        y_test,
        pred,
        target_names=le.classes_
    )
)

print(
    confusion_matrix(
        y_test,
        pred
    )
)

# ==========================================
# Test Similar Words
# ==========================================

print(
    ft_model.wv.most_similar(
        "python"
    )
)

# ==========================================
# Test OOV Word
# ==========================================

print(
    ft_model.wv["pythonnnn"].shape
)

print(
    ft_model.wv["tensorflowww"].shape
)


Accuracy : 0.8287795992714025
                        precision    recall  f1-score   support

Database_Administrator       0.89      0.85      0.87       184
   Front_End_Developer       0.75      0.61      0.67       106
        Java_Developer       0.94      0.94      0.94       249
 Network_Administrator       0.77      0.59      0.67        94
       Project_manager       0.82      0.92      0.87       144
      Python_Developer       0.97      0.97      0.97       256
      Security_Analyst       0.91      0.87      0.89       175
    Software_Developer       0.59      0.26      0.37        87
 Systems_Administrator       0.73      0.85      0.79       155
         Web_Developer       0.64      0.83      0.72       197

              accuracy                           0.83      1647
             macro avg       0.80      0.77      0.78      1647
          weighted avg       0.83      0.83      0.82      1647

[[156   0   1   2  16   0   5   0   4   0]
 [  0  65   2   0   0   0   

# Resume Classification - NLP Model Comparison

## Dataset

- Multi-class Resume Classification Dataset
- Classes:
  - Python Developer
  - Java Developer
  - Web Developer
  - Project Manager
  - Database Administrator
  - Security Analyst
  - Network Administrator
  - Systems Administrator
  - Software Developer
  - Front-End Developer

---

# Models Evaluated

| Model | Accuracy |
|---------|---------:|
| Word2Vec + GaussianNB | 71.8% |
| Word2Vec + Logistic Regression | 82.4% |
| FastText + Logistic Regression | 82.9% |
| TF-IDF + LinearSVC | 93.4% |

---

# Key Findings

## 1. TF-IDF + LinearSVC performed best

Accuracy:

```text
93.4%
```

Reasons:

- Resume classification is heavily keyword-driven.
- Important technologies such as:
  - Python
  - TensorFlow
  - AWS
  - Docker
  - Spring
  - Kubernetes

  are preserved as individual features.

- LinearSVC works exceptionally well on:
  - High-dimensional
  - Sparse
  - TF-IDF feature spaces

---

## 2. Word2Vec improved significantly with Logistic Regression

Accuracy:

```text
71.8% → 82.4%
```

Comparison:

| Model | Accuracy |
|---------|---------:|
| Word2Vec + GaussianNB | 71.8% |
| Word2Vec + Logistic Regression | 82.4% |

Reason:

- GaussianNB assumes feature independence and Gaussian distributions.
- Word embeddings contain correlated continuous features.
- Logistic Regression handles embedding vectors more effectively.

---

## 3. FastText slightly outperformed Word2Vec

Accuracy:

```text
Word2Vec : 82.4%
FastText : 82.9%
```

Improvement:

```text
≈ 0.5%
```

Reason:

- FastText uses character n-grams.
- Can generate embeddings for rare or unseen words.
- Better handling of:
  - Python2
  - Python27
  - Python3x
  - Misspellings
  - Rare skills

---

## 4. FastText Learned Subword Relationships

Most similar words to:

```text
python
```

included:

```text
python2
python27
python3x
modpython
rubypython
javapython
```

Inference:

- FastText groups words using character structure.
- Demonstrates successful subword learning.

---

## 5. Word2Vec Learned Context-Based Relationships

Most similar words to:

```text
python
```

included:

```text
perl
lua
erlang
```

Inference:

- Word2Vec groups words appearing in similar contexts.
- Learns semantic similarity rather than character similarity.

---

# Class-wise Performance Analysis

## Best Performing Class

### Python Developer

```text
Precision : 0.97
Recall    : 0.97
F1-Score  : 0.97
```

Reason:

- Strongly identifiable keywords:
  - Python
  - Django
  - Flask
  - Pandas
  - NumPy

---

## Weakest Performing Class

### Software Developer

```text
Precision : 0.59
Recall    : 0.26
```

Reason:

- Generic role description.
- Significant overlap with:
  - Web Developer
  - Python Developer
  - Java Developer

- Class boundaries are less distinct.

---

# Confusion Matrix Insights

Most common misclassifications:

```text
Software Developer → Web Developer
Network Administrator → Systems Administrator
Front-End Developer → Web Developer
```

Reason:

- Similar technologies and skills appear across these roles.
- Resume content overlaps considerably.

---

# Overall Conclusion

TF-IDF + LinearSVC remains the strongest model for this resume classification dataset because it preserves important keywords explicitly.

Word2Vec and FastText successfully learn semantic and subword relationships but lose some discriminative information when document embeddings are created using simple averaging.

The experiment demonstrates that:

- Better embeddings do not always lead to better classification accuracy.
- For keyword-heavy datasets, TF-IDF remains a highly competitive baseline.
- FastText provides better handling of rare and unseen words compared to Word2Vec.
- Model selection should depend on the nature of the dataset rather than assuming newer embeddings will always perform better.